diff --git a/setup.py b/setup.py
index dd87f7d..db9f8aa 100644
--- a/setup.py
+++ b/setup.py
@@ -63,8 +63,7 @@ CLASSIFIERS = [
     "Programming Language :: Python :: 3.8",
 ]
 
-version_txt = os.path.join(here, 'supervisor/version.txt')
-supervisor_version = open(version_txt).read().strip()
+supervisor_version = "4.2.0"
 
 dist = setup(
     name='supervisor',
diff --git a/supervisor/options.py b/supervisor/options.py
index 3dc6024..2886d17 100644
--- a/supervisor/options.py
+++ b/supervisor/options.py
@@ -54,12 +54,7 @@ from supervisor import states
 from supervisor import xmlrpc
 from supervisor import poller
 
-def _read_version_txt():
-    mydir = os.path.abspath(os.path.dirname(__file__))
-    version_txt = os.path.join(mydir, 'version.txt')
-    with open(version_txt, 'r') as f:
-        return f.read().strip()
-VERSION = _read_version_txt()
+VERSION = "4.2.0"
 
 def normalize_path(v):
     return os.path.normpath(os.path.abspath(os.path.expanduser(v)))
